File: client/src/lib/announcementSeen.ts
const STORAGE_PREFIX = 'openmusic_announcement_seen:';

export function getAnnouncementRevision(enabled: boolean, text: string): string {
  return `${enabled ? 1 : 0}\n${text.trim()}`;
}

function storageKey(roomId: string): string {
  return `${STORAGE_PREFIX}${roomId}`;
}

export function hasSeenAnnouncement(roomId: string, enabled: boolean, text: string): boolean {
  try {
    return localStorage.getItem(storageKey(roomId)) === getAnnouncementRevision(enabled, text);
  } catch {
    return false;
  }
}

export function markAnnouncementSeen(roomId: string, enabled: boolean, text: string): void {
  try {
    localStorage.setItem(storageKey(roomId), getAnnouncementRevision(enabled, text));
  } catch {
    // localStorage may be unavailable.
  }
}

export function shouldAutoShowAnnouncement(
  roomId: string,
  enabled?: boolean,
  text?: string,
): boolean {
  if (!enabled || !text?.trim()) return false;
  return !hasSeenAnnouncement(roomId, enabled, text);
}

File: mobile/lib/core/announcement_seen.dart
import 'package:shared_preferences/shared_preferences.dart';

const _prefix = 'openmusic_announcement_seen:';

String announcementRevision(bool enabled, String text) =>
    '${enabled ? 1 : 0}\n${text.trim()}';

Future<bool> shouldAutoShowAnnouncement({
  required String roomId,
  required bool enabled,
  String? text,
}) async {
  final body = text?.trim() ?? '';
  if (!enabled || body.isEmpty) return false;
  final prefs = await SharedPreferences.getInstance();
  return prefs.getString('$_prefix$roomId') != announcementRevision(enabled, body);
}

Future<void> markAnnouncementSeen({
  required String roomId,
  required bool enabled,
  String? text,
}) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.setString(
    '$_prefix$roomId',
    announcementRevision(enabled, text ?? ''),
  );
}
